### ISSUE TITLE
[STORM-839] Deadlock hazard in backtype.storm.messaging.netty.Client

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#4679]
<p>See the thread dump below that shows the deadlock. client-worker-1 is holding 7b5a7fa5 and waiting on 1446a1e9. Thread-10-disruptor-worker-transfer-queue is holding 1446a1e9 and is waiting on 7b5a7fa5.</p>

<p>(Thread dump is truncated to show only the relevant parts)</p>

<p>2015-05-28 15:37:15<br/>
Full thread dump Java HotSpot(TM) 64-Bit Server VM (24.72-b04 mixed mode):</p>

<p>"Thread-10-disruptor-worker-transfer-queue" - Thread t@52<br/>
   java.lang.Thread.State: BLOCKED<br/>
	at org.apache.storm.netty.channel.socket.nio.AbstractNioWorker.cleanUpWriteBuffer(AbstractNioWorker.java:398)</p>
<ul class="alternate" type="square">
	<li>waiting to lock <7b5a7fa5> (a java.lang.Object) owned by "client-worker-1" t@25<br/>
	at org.apache.storm.netty.channel.socket.nio.AbstractNioWorker.writeFromUserCode(AbstractNioWorker.java:128)<br/>
	at org.apache.storm.netty.channel.socket.nio.NioClientSocketPipelineSink.eventSunk(NioClientSocketPipelineSink.java:84)<br/>
	at org.apache.storm.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendDownstream(DefaultChannelPipeline.java:779)<br/>
	at org.apache.storm.netty.channel.Channels.write(Channels.java:725)<br/>
	at org.apache.storm.netty.handler.codec.oneone.OneToOneEncoder.doEncode(OneToOneEncoder.java:71)<br/>
	at org.apache.storm.netty.handler.codec.oneone.OneToOneEncoder.handleDownstream(OneToOneEncoder.java:59)<br/>
	at org.apache.storm.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:591)<br/>
	at org.apache.storm.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:582)<br/>
	at org.apache.storm.netty.channel.Channels.write(Channels.java:704)<br/>
	at org.apache.storm.netty.channel.Channels.write(Channels.java:671)<br/>
	at org.apache.storm.netty.channel.AbstractChannel.write(AbstractChannel.java:248)<br/>
	at backtype.storm.messaging.netty.Client.flushMessages(Client.java:480)</li>
	<li>locked <1446a1e9> (a backtype.storm.messaging.netty.Client)<br/>
	at backtype.storm.messaging.netty.Client.send(Client.java:412)</li>
	<li>locked <1446a1e9> (a backtype.storm.messaging.netty.Client)<br/>
	at backtype.storm.utils.TransferDrainer.send(TransferDrainer.java:54)<br/>
	at backtype.storm.daemon.worker$mk_transfer_tuples_handler$fn_<em>5014$fn</em>_5015.invoke(worker.clj:334)<br/>
	at backtype.storm.daemon.worker$mk_transfer_tuples_handler$fn__5014.invoke(worker.clj:332)<br/>
	at backtype.storm.disruptor$clojure_handler$reify__1446.onEvent(disruptor.clj:58)<br/>
	at backtype.storm.utils.DisruptorQueue.consumeBatchToCursor(DisruptorQueue.java:125)<br/>
	at backtype.storm.utils.DisruptorQueue.consumeBatchWhenAvailable(DisruptorQueue.java:99)<br/>
	at backtype.storm.disruptor$consume_batch_when_available.invoke(disruptor.clj:80)<br/>
	at backtype.storm.disruptor$consume_loop_STAR_$fn__1459.invoke(disruptor.clj:94)<br/>
	at backtype.storm.util$async_loop$fn__458.invoke(util.clj:463)<br/>
	at clojure.lang.AFn.run(AFn.java:24)<br/>
	at java.lang.Thread.run(Unknown Source)</li>
</ul>

<p>   Locked ownable synchronizers:</p>
<ul class="alternate" type="square">
	<li>None</li>
</ul>

<p>"client-worker-1" - Thread t@25<br/>
   java.lang.Thread.State: BLOCKED<br/>
	at backtype.storm.messaging.netty.Client.closeChannelAndReconnect(Client.java:501)</p>
<ul class="alternate" type="square">
	<li>waiting to lock <1446a1e9> (a backtype.storm.messaging.netty.Client) owned by "Thread-10-disruptor-worker-transfer-queue" t@52<br/>
	at backtype.storm.messaging.netty.Client.access$1400(Client.java:78)<br/>
	at backtype.storm.messaging.netty.Client$3.operationComplete(Client.java:492)<br/>
	at org.apache.storm.netty.channel.DefaultChannelFuture.notifyListener(DefaultChannelFuture.java:427)<br/>
	at org.apache.storm.netty.channel.DefaultChannelFuture.notifyListeners(DefaultChannelFuture.java:413)<br/>
	at org.apache.storm.netty.channel.DefaultChannelFuture.setFailure(DefaultChannelFuture.java:380)<br/>
	at org.apache.storm.netty.channel.socket.nio.AbstractNioWorker.cleanUpWriteBuffer(AbstractNioWorker.java:437)</li>
	<li>locked <7b5a7fa5> (a java.lang.Object)<br/>
	at org.apache.storm.netty.channel.socket.nio.AbstractNioWorker.close(AbstractNioWorker.java:373)<br/>
	at org.apache.storm.netty.channel.socket.nio.NioWorker.read(NioWorker.java:93)<br/>
	at org.apache.storm.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)<br/>
	at org.apache.storm.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:318)<br/>
	at org.apache.storm.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)<br/>
	at org.apache.storm.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)<br/>
	at org.apache.storm.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)<br/>
	at org.apache.storm.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)<br/>
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)<br/>
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)<br/>
	at java.lang.Thread.run(Unknown Source)</li>
</ul>

<p>   Locked ownable synchronizers:</p>
<ul class="alternate" type="square">
	<li>locked <75e528fd> (a java.util.concurrent.ThreadPoolExecutor$Worker)</li>
</ul>

---
<details><summary><i>Originally reported by <a title="Enno Shioji" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=eshioji">eshioji</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-839" target="_blank">Deadlock hazard in backtype.storm.messaging.netty.Client</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Enno Shioji" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=eshioji">eshioji</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Critical
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2015-07-14T20:19:06+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=eshioji">eshioji</a>:</i>
<p>I've started working on a fix because the topology I'm working on is behaving unstable, which I'm suspecting is because of this. I'll submit a pull request when the code is tested.</p>

<p>Branch is at: <a href="https://github.com/eshioji/storm/tree/STORM-839" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/eshioji/storm/tree/STORM-839</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user eshioji opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/565" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/565</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-839" title="Deadlock hazard in backtype.storm.messaging.netty.Client" class="issue-link" data-issue-key="STORM-839"><del>STORM-839</del></a></p>

<p>    This fixes the reported deadlock between `disruptor-worker-transfer-queue` thread and `client-worker` thread, which seem to have been introduced by <a href="https://issues.apache.org/jira/browse/STORM-329" title="Fix cascading Storm failure by improving reconnection strategy and buffering messages" class="issue-link" data-issue-key="STORM-329"><del>STORM-329</del></a>.</p>

<p>    After reviewing the `v0.9.4` code, my conclusion was that the background flushing task can be removed without real change in the behavior. By doing this, the synchronization that was involved in the deadlock can be removed altogether.</p>

<p>    My reasoning is as follows:</p>
<ul class="alternate" type="square">
	<li>One has three option to deal with Netty's buffer filling up:<br/>
      1. Discard incoming new messages<br/>
      2. Block client thread until there is space (back pressure)<br/>
      3. Keep buffering up until OOME is thrown</li>
	<li>My guess is that the `v0.9.4` code attempted to implement option <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/information.png" height="16" width="16" align="absmiddle" alt="" border="0"/>, but actually the behavior is option (iii). When Netty's `Channel.isWritable` returns false in `Client.send`, the thread avoids writing to the `Channel` and leaves messages in the `Channel.messageBatch` field, which the background task flushes when the `Channel.isWritable` starts to return true again. However, if `Client.send` is called again before this (which should be common), the content of `Channel.messageBatch` is written and buffered on the `Channel` anyways, because `flushMessages` does not check `Channel.isWritable`.</li>
	<li>AFAIU we like option (ii), but this requires more work. Between option <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/information.png" height="16" width="16" align="absmiddle" alt="" border="0"/> and option (iii), IMO (iii) is superior because if OOME happens, the user can reduce MAX_PENDING_TUPLES to prevent this. Discarding messages would be harder to diagnose.</li>
	<li>If we are content with option (ii), we can remove the background flushing task and related synchronization altogether, removing the source of deadlock. We'd simply write and buffer onto the unbounded buffer of `Channel`, and if there are too many pending messages, the worker will die of OOME, and the user should reduce this with indirect means like reducing MAX_PENDING_TUPLES until option (ii) is implemented in the future.</li>
</ul>


<p>    Thoughts @miguno ?</p>





<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/eshioji/storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/eshioji/storm</a> <a href="https://issues.apache.org/jira/browse/STORM-839" title="Deadlock hazard in backtype.storm.messaging.netty.Client" class="issue-link" data-issue-key="STORM-839"><del>STORM-839</del></a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/565.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/565.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #565</p>

<hr />
<p>commit 98f4e619d54052b73a309d23ab7214953e4c7774<br/>
Author: Sriharsha Chintalapani <mail@harsha.io><br/>
Date:   2015-02-05T18:08:05Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-130" title="[Storm 0.8.2]: java.io.FileNotFoundException: File &#39;../stormconf.ser&#39; does not exist" class="issue-link" data-issue-key="STORM-130"><del>STORM-130</del></a>: Supervisor getting killed due to java.io.FileNotFoundException: File '../stormconf.ser' does not exist.</p>

<p>    Signed-off-by: P. Taylor Goetz <ptgoetz@gmail.com></p>

<p>commit a1e5893e1b94c224d39fedf11583b216c21351c8<br/>
Author: P. Taylor Goetz <ptgoetz@gmail.com><br/>
Date:   2015-02-24T20:46:12Z</p>

<p>    update changelog for <a href="https://issues.apache.org/jira/browse/STORM-130" title="[Storm 0.8.2]: java.io.FileNotFoundException: File &#39;../stormconf.ser&#39; does not exist" class="issue-link" data-issue-key="STORM-130"><del>STORM-130</del></a></p>

<p>commit 62788f295bb1fb1cc83b99c30f82beb40eea5f25<br/>
Author: P. Taylor Goetz <ptgoetz@gmail.com><br/>
Date:   2015-02-24T23:03:40Z</p>

<p>    port <a href="https://issues.apache.org/jira/browse/STORM-329" title="Fix cascading Storm failure by improving reconnection strategy and buffering messages" class="issue-link" data-issue-key="STORM-329"><del>STORM-329</del></a> fix to 0.9.x</p>

<p>commit 81016c2ed7222da99138bc9971e335533d4cb518<br/>
Author: Michael G. Noll <mnoll@verisign.com><br/>
Date:   2015-02-16T09:01:27Z</p>

<p>    Track how many messages are being dropped when a connection is unavailable</p>

<p>    Signed-off-by: P. Taylor Goetz <ptgoetz@gmail.com></p>

<p>commit 97a76fc896de508f015dbe32f1473ddbf10d736b<br/>
Author: Michael G. Noll <mnoll@verisign.com><br/>
Date:   2015-02-16T09:03:07Z</p>

<p>    Clarify name of method for dropping messages</p>

<p>    Signed-off-by: P. Taylor Goetz <ptgoetz@gmail.com></p>

<p>commit 9138d9fc255639b4d0d43657379ce467591e8ef2<br/>
Author: Michael G. Noll <mnoll@verisign.com><br/>
Date:   2015-02-16T09:07:35Z</p>

<p>    Change log level for intentionally dropping messages from WARN to ERROR</p>

<p>    This change makes the log level for dropping messages consistent in<br/>
    Client.java.</p>

<p>    Signed-off-by: P. Taylor Goetz <ptgoetz@gmail.com></p>

<p>commit 6b06d8468ff5e743fb12b85dd84fe0931041c2c3<br/>
Author: P. Taylor Goetz <ptgoetz@gmail.com><br/>
Date:   2015-02-24T23:18:43Z</p>

<p>    add <a href="https://issues.apache.org/jira/browse/STORM-329" title="Fix cascading Storm failure by improving reconnection strategy and buffering messages" class="issue-link" data-issue-key="STORM-329"><del>STORM-329</del></a> to changelog</p>

<p>commit e63fb2af9086e2b2e688662ca42a4b4d0112274b<br/>
Author: Parth Brahmbhatt <brahmbhatt.parth@gmail.com><br/>
Date:   2015-03-03T00:06:58Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-693" title="KafkaBolt exception handling improvement" class="issue-link" data-issue-key="STORM-693"><del>STORM-693</del></a>: when bolt fails to write tuple, it should report error instead of silently acking.</p>

<p>    Signed-off-by: P. Taylor Goetz <ptgoetz@gmail.com></p>

<p>commit 92836de540ec8ab90d7591b96ba02126e80b5c3a<br/>
Author: P. Taylor Goetz <ptgoetz@gmail.com><br/>
Date:   2015-03-18T14:59:56Z</p>

<p>    add <a href="https://issues.apache.org/jira/browse/STORM-693" title="KafkaBolt exception handling improvement" class="issue-link" data-issue-key="STORM-693"><del>STORM-693</del></a> to changelog</p>

<p>commit c19e482b70f18d690ad165c78551860506486095<br/>
Author: Parth Brahmbhatt <brahmbhatt.parth@gmail.com><br/>
Date:   2015-02-20T19:56:22Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-682" title="Supervisor local worker state corrupted and failing to start." class="issue-link" data-issue-key="STORM-682"><del>STORM-682</del></a>: supervisor should handle worker state corruption gracefully.</p>

<p>    Signed-off-by: P. Taylor Goetz <ptgoetz@gmail.com></p>

<p>commit f0de11a20fe2f20dc1dc2f485549e0dc342f8680<br/>
Author: P. Taylor Goetz <ptgoetz@gmail.com><br/>
Date:   2015-03-18T15:05:30Z</p>

<p>    add <a href="https://issues.apache.org/jira/browse/STORM-682" title="Supervisor local worker state corrupted and failing to start." class="issue-link" data-issue-key="STORM-682"><del>STORM-682</del></a> to changelog</p>

<p>commit 835a410c879dc1eb02d9670410f65fe0be6f28c6<br/>
Author: Parth Brahmbhatt <brahmbhatt.parth@gmail.com><br/>
Date:   2015-01-14T20:27:35Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-559" title="ZkHosts in README should use 2181 as port" class="issue-link" data-issue-key="STORM-559"><del>STORM-559</del></a>: ZkHosts in README should use 2181 as port.</p>

<p>    Signed-off-by: P. Taylor Goetz <ptgoetz@gmail.com></p>

<p>commit 30e0be8616c89cb1f8a51fcf462f76a075e6e964<br/>
Author: P. Taylor Goetz <ptgoetz@gmail.com><br/>
Date:   2015-03-18T15:11:16Z</p>

<p>    add <a href="https://issues.apache.org/jira/browse/STORM-559" title="ZkHosts in README should use 2181 as port" class="issue-link" data-issue-key="STORM-559"><del>STORM-559</del></a> to changelog</p>

<p>commit b1bbacb7134d17ff47c2e8b8857a66244a4d1d4f<br/>
Author: P. Taylor Goetz <ptgoetz@gmail.com><br/>
Date:   2015-03-18T15:28:11Z</p>

<p>    add missing import in supervisor.clj</p>

<p>commit edf596bac8feab0c8721f7de94474e5549858355<br/>
Author: P. Taylor Goetz <ptgoetz@gmail.com><br/>
Date:   2015-03-18T16:21:38Z</p>

<p>    <span class="error">&#91;maven-release-plugin&#93;</span> prepare release v0.9.4</p>

<p>commit 48d10e20eb3c750fc41fcf0bef3d49501cf6d5a4<br/>
Author: P. Taylor Goetz <ptgoetz@gmail.com><br/>
Date:   2015-03-18T16:21:45Z</p>

<p>    <span class="error">&#91;maven-release-plugin&#93;</span> prepare for next development iteration</p>

<p>commit 41f44f9914d4f27d0db3f211a85f88301533f09b<br/>
Author: P. Taylor Goetz <ptgoetz@gmail.com><br/>
Date:   2015-03-18T17:59:39Z</p>

<p>    Revert "<span class="error">&#91;maven-release-plugin&#93;</span> prepare for next development iteration"</p>

<p>    This reverts commit 48d10e20eb3c750fc41fcf0bef3d49501cf6d5a4.</p>

<p>commit 233603c3cbd729fdfabd2759bfa7705811996aa4<br/>
Author: P. Taylor Goetz <ptgoetz@gmail.com><br/>
Date:   2015-03-18T18:00:11Z</p>

<p>    Revert "<span class="error">&#91;maven-release-plugin&#93;</span> prepare release v0.9.4"</p>

<p>    This reverts commit edf596bac8feab0c8721f7de94474e5549858355.</p>

<p>commit 61e1b5c3e226122143c91bbd7527b605f8ed8727<br/>
Author: P. Taylor Goetz <ptgoetz@gmail.com><br/>
Date:   2015-03-18T18:05:51Z</p>

<p>    add Apache license header to ConnectionWithStatus.java</p>

<p>commit 00091d7952681a39281aa171adfad133a5e26330<br/>
Author: P. Taylor Goetz <ptgoetz@gmail.com><br/>
Date:   2015-03-18T18:13:57Z</p>

<p>    <span class="error">&#91;maven-release-plugin&#93;</span> prepare release v0.9.4</p>

<p>commit ed8ab3ec194f19c75fc2f5c000609204f04b50e8<br/>
Author: Enno Shioji <eshioji@gmail.com><br/>
Date:   2015-05-28T19:42:05Z</p>

<p>    Simplified the flow and removed the lock that was causing the deadlock</p>

<p>commit 91b8eb3840432e47b79f40abebec8304627732a8<br/>
Author: Enno Shioji <eshioji@gmail.com><br/>
Date:   2015-05-28T19:46:17Z</p>

<p>    Bump version</p>

<p>commit b7d84bdc7fd3de34f45a94131cdbb6bfbd3763dc<br/>
Author: Enno Shioji <eshioji@gmail.com><br/>
Date:   2015-05-28T21:27:31Z</p>

<p>    Remove background flushing because it doesn't seem necessary. Netty's Channel queues up written data on an unbounded buffer. The background flushing seems to have been added to avoid this, but in practice it was probably doing it anyways because flushMessages(), which is called by send() doesn't check for isWritable. Moreover, queuing on an unbounded buffer seems fine because back pressure is provided by MAX_PENDING_TUPLE. If OOME occurs due to this buffer overflowing, it seems reasonable that one has to reduce MAX_PENDING_TUPLE, rather than Storm trying to cope with it by dropping messages.</p>

<p>commit 679e42bc1e38f51c2759667b03cb45322c6a793b<br/>
Author: Enno Shioji <eshioji@gmail.com><br/>
Date:   2015-05-28T21:31:35Z</p>

<p>    Change to a SNAPSHOT version for deployment purposes</p>

<p>commit 27a92e2aa3488c0203f500306e0583ff9e7e1e82<br/>
Author: Enno Shioji <eshioji@gmail.com><br/>
Date:   2015-05-29T09:32:16Z</p>

<p>    Remove (now) dead comment and code</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user eshioji commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/565#issuecomment-106765335" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/565#issuecomment-106765335</a></p>

<p>    Argh sorry I should have opened this against Storm v0.9.4, I worked off v0.9.4 tag.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user eshioji closed the pull request at:</p>

<p>    <a href="https://github.com/apache/storm/pull/565" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/565</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user eshioji opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/566" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/566</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-839" title="Deadlock hazard in backtype.storm.messaging.netty.Client" class="issue-link" data-issue-key="STORM-839"><del>STORM-839</del></a> Deadlock hazard in backtype.storm.messaging.netty.Client</p>

<p>    (I accidentally did a PR against master with the same content, please ignore that one)</p>

<p>    This fixes the reported deadlock between `disruptor-worker-transfer-queue` thread and `client-worker` thread, which seem to have been introduced by <a href="https://issues.apache.org/jira/browse/STORM-329" title="Fix cascading Storm failure by improving reconnection strategy and buffering messages" class="issue-link" data-issue-key="STORM-329"><del>STORM-329</del></a>.</p>

<p>    After reviewing the `v0.9.4` code, my conclusion was that the background flushing task can be removed without real change in the behavior. By doing this, the synchronization that was involved in the deadlock can be removed altogether.</p>

<p>    My reasoning is as follows:</p>
<ul class="alternate" type="square">
	<li>One has three option to deal with Netty's buffer filling up:<br/>
      1. Discard incoming new messages<br/>
      2. Block client thread until there is space (back pressure)<br/>
      3. Keep buffering up until OOME is thrown</li>
	<li>My guess is that the `v0.9.4` code attempted to implement option <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/information.png" height="16" width="16" align="absmiddle" alt="" border="0"/>, but actually the behavior is option (iii). When Netty's `Channel.isWritable` returns false in `Client.send`, the thread avoids writing to the `Channel` and leaves messages in the `Channel.messageBatch` field, which the background task flushes when the `Channel.isWritable` starts to return true again. However, if `Client.send` is called again before this (which should be common), the content of `Channel.messageBatch` is written and buffered on the `Channel` anyways, because `flushMessages` does not check `Channel.isWritable`.</li>
	<li>AFAIU we like option (ii), but this requires more work. Between option <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/information.png" height="16" width="16" align="absmiddle" alt="" border="0"/> and option (iii), IMO (iii) is superior because if OOME happens, the user can reduce MAX_PENDING_TUPLES to prevent this. Discarding messages would be harder to diagnose.</li>
	<li>If we are content with option (ii), we can remove the background flushing task and related synchronization altogether, removing the source of deadlock. We'd simply write and buffer onto the unbounded buffer of `Channel`, and if there are too many pending messages, the worker will die of OOME, and the user should reduce this with indirect means like reducing MAX_PENDING_TUPLES until option (ii) is implemented in the future.</li>
</ul>


<p>    Thoughts @miguno ?</p>





<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/eshioji/storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/eshioji/storm</a> <a href="https://issues.apache.org/jira/browse/STORM-839" title="Deadlock hazard in backtype.storm.messaging.netty.Client" class="issue-link" data-issue-key="STORM-839"><del>STORM-839</del></a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/566.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/566.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #566</p>

<hr />
<p>commit ed8ab3ec194f19c75fc2f5c000609204f04b50e8<br/>
Author: Enno Shioji <eshioji@gmail.com><br/>
Date:   2015-05-28T19:42:05Z</p>

<p>    Simplified the flow and removed the lock that was causing the deadlock</p>

<p>commit 91b8eb3840432e47b79f40abebec8304627732a8<br/>
Author: Enno Shioji <eshioji@gmail.com><br/>
Date:   2015-05-28T19:46:17Z</p>

<p>    Bump version</p>

<p>commit b7d84bdc7fd3de34f45a94131cdbb6bfbd3763dc<br/>
Author: Enno Shioji <eshioji@gmail.com><br/>
Date:   2015-05-28T21:27:31Z</p>

<p>    Remove background flushing because it doesn't seem necessary. Netty's Channel queues up written data on an unbounded buffer. The background flushing seems to have been added to avoid this, but in practice it was probably doing it anyways because flushMessages(), which is called by send() doesn't check for isWritable. Moreover, queuing on an unbounded buffer seems fine because back pressure is provided by MAX_PENDING_TUPLE. If OOME occurs due to this buffer overflowing, it seems reasonable that one has to reduce MAX_PENDING_TUPLE, rather than Storm trying to cope with it by dropping messages.</p>

<p>commit 679e42bc1e38f51c2759667b03cb45322c6a793b<br/>
Author: Enno Shioji <eshioji@gmail.com><br/>
Date:   2015-05-28T21:31:35Z</p>

<p>    Change to a SNAPSHOT version for deployment purposes</p>

<p>commit 27a92e2aa3488c0203f500306e0583ff9e7e1e82<br/>
Author: Enno Shioji <eshioji@gmail.com><br/>
Date:   2015-05-29T09:32:16Z</p>

<p>    Remove (now) dead comment and code</p>

<p>commit 09bf6e1b5d9d351f2a60cd9a32e0239752cf437a<br/>
Author: Enno Shioji <eshioji@gmail.com><br/>
Date:   2015-05-29T10:23:46Z</p>

<p>    Merge branch '0.9.x-branch' into <a href="https://issues.apache.org/jira/browse/STORM-839" title="Deadlock hazard in backtype.storm.messaging.netty.Client" class="issue-link" data-issue-key="STORM-839"><del>STORM-839</del></a></p>

<p>    Conflicts:<br/>
    	examples/storm-starter/pom.xml<br/>
    	external/storm-hbase/pom.xml<br/>
    	external/storm-hdfs/pom.xml<br/>
    	external/storm-kafka/pom.xml<br/>
    	pom.xml<br/>
    	storm-buildtools/maven-shade-clojure-transformer/pom.xml<br/>
    	storm-core/pom.xml<br/>
    	storm-dist/binary/pom.xml<br/>
    	storm-dist/source/pom.xml</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user miguno commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/566#issuecomment-106795826" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/566#issuecomment-106795826</a></p>

<p>    > My reasoning is as follows:<br/>
    > <br/>
    > * One has three option to deal with Netty's buffer filling up:<br/>
    >     1. Discard incoming new messages<br/>
    >     2. Block client thread until there is space (back pressure)<br/>
    >     3. Keep buffering up until OOME is thrown<br/>
    ><br/>
    > My guess is that the v0.9.4 code attempted to implement option <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/information.png" height="16" width="16" align="absmiddle" alt="" border="0"/>, but actually the behavior is option (iii).</p>

<p>    The code's intention was actually (iii).  As you described back pressure (ii) was not picked because this will require a significant amount of work, which was thus out of scope for fixing <a href="https://issues.apache.org/jira/browse/STORM-329" title="Fix cascading Storm failure by improving reconnection strategy and buffering messages" class="issue-link" data-issue-key="STORM-329"><del>STORM-329</del></a>.  The reason (iii) was preferred over <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/information.png" height="16" width="16" align="absmiddle" alt="" border="0"/> was also as you described &#8211; if (and only if) users have enabled acking = guaranteed message processing for a topology, they can prevent OOM errors from happening by setting <span class="error">&#91;`topology.max.spout.pending`&#93;</span>(<a href="https://github.com/apache/storm/blob/master/conf/defaults.yaml#L180" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/blob/master/conf/defaults.yaml#L180</a>).</p>

<p>    I'll have to look at your code in more detail before commenting.  I'll also ping @ptgoetz, @clockfly, @tedxia, and @revans2 who were involved in this significant patch (sorry for the spam, folks!).</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user eshioji commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/566#issuecomment-106800904" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/566#issuecomment-106800904</a></p>

<p>    Just as a heads up, I'm seeing <a href="https://issues.apache.org/jira/browse/STORM-763" title="nimbus reassigned worker A to another machine, but other worker&#39;s netty client can&#39;t connect to the new worker A " class="issue-link" data-issue-key="STORM-763"><del>STORM-763</del></a> as well (300 ERROR messages per second per machine) in our cluster. My initial gut feeling is that maybe reconnect isn't getting enough run time from the ScheduledExecutor. Depending on the results I might create another PR for 763.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user eshioji opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/616" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/616</a></p>

<p>    Storm 763/839 0.11.x</p>

<p>    This is a port of PR #568 . It fixes <a href="https://issues.apache.org/jira/browse/STORM-839" title="Deadlock hazard in backtype.storm.messaging.netty.Client" class="issue-link" data-issue-key="STORM-839"><del>STORM-839</del></a> (Deadlock) and <a href="https://issues.apache.org/jira/browse/STORM-763" title="nimbus reassigned worker A to another machine, but other worker&#39;s netty client can&#39;t connect to the new worker A " class="issue-link" data-issue-key="STORM-763"><del>STORM-763</del></a> (Establish Netty reconnects asynchronously and reduce verbosity of error logs)</p>

<p>    @revans2 This is the PR for master. I ran the performance test against master's HEAD and the results were in-line with the results I got in #568 . I'll create a separate PR for 0.10.x</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/eshioji/storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/eshioji/storm</a> <a href="https://issues.apache.org/jira/browse/STORM-763" title="nimbus reassigned worker A to another machine, but other worker&#39;s netty client can&#39;t connect to the new worker A " class="issue-link" data-issue-key="STORM-763"><del>STORM-763</del></a>_0.11.x</p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/616.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/616.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #616</p>

<hr />
<p>commit b1411aa63383801913cb1340a8b51c5bb46db0ba<br/>
Author: Enno Shioji <eshioji@gmail.com><br/>
Date:   2015-07-01T12:54:33Z</p>

<p>    This fixes <a href="https://issues.apache.org/jira/browse/STORM-763" title="nimbus reassigned worker A to another machine, but other worker&#39;s netty client can&#39;t connect to the new worker A " class="issue-link" data-issue-key="STORM-763"><del>STORM-763</del></a> and <a href="https://issues.apache.org/jira/browse/STORM-839" title="Deadlock hazard in backtype.storm.messaging.netty.Client" class="issue-link" data-issue-key="STORM-839"><del>STORM-839</del></a></p>

<p>commit 935e87accc85340549f44effa8a675e950747faa<br/>
Author: Enno Shioji <eshioji@gmail.com><br/>
Date:   2015-07-02T17:56:23Z</p>

<p>    Remove obsolete TODO</p>

<p>commit 7af467723b5c4b0beedc40626a8c56bc7c3e0d21<br/>
Author: Enno Shioji <eshioji@gmail.com><br/>
Date:   2015-07-03T00:02:00Z</p>

<p>    Bring back removal of client from context upon closing</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/storm/pull/617" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/617</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/storm/pull/566" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/566</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p>Thanks <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=eshioji" class="user-hover" rel="eshioji">Enno Shioji</a>,</p>

<p>I merged this into master, branch-0.10.x and branch-0.9.x.  You may want to talk to <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ptgoetz" class="user-hover" rel="ptgoetz">P. Taylor Goetz</a> about doing a 0.9.6 release.</p>
